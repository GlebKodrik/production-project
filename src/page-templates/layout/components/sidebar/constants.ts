import { ROUTES_PATH } from 'constants/routers';
import { ICON_NAMES } from 'shared-components/icon/constants/icon-names';
import { TRoutesPath } from 'types/routes';
import { TIconNames } from 'shared-components/icon';

type TLinks = {
  name: string,
  link: TRoutesPath,
  icon: TIconNames
};

export const LINKS: TLinks[] = [
  { name: 'sidebar.links.aboutLink', link: ROUTES_PATH.ABOUT, icon: ICON_NAMES.CLARITY_LIST },
  { name: 'sidebar.links.mainLink', link: ROUTES_PATH.BASE, icon: ICON_NAMES.HOME },
  { name: 'sidebar.links.profile', link: ROUTES_PATH.PROFILE, icon: ICON_NAMES.PROFILE },
];
