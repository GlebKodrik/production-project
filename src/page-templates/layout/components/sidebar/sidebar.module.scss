.sidebar {
  position: relative;
  width: var(--sidebar-width);
  height: calc(100vh - var(--navbar-height));
  background-color: var(--inverted-bg-color);
  transition: width 0.4s;
}

.link-wrapper {
  display: flex;
  flex-direction: column;
  margin-top: 20px;
  margin-left: 30px;
  overflow: hidden;

  .item {
    display: flex;
    align-items: center;
    width: var(--sidebar-width);

    &:first-child {
      margin-bottom: 10px;
    }

    .icon {
      fill: var(--inverted-primary-color);
    }
  }

  .link-text {
    margin-left: 5px;
    opacity: 1;
  }
}

.switchers {
  position: absolute;
  bottom: 20px;
  display: flex;
  justify-content: center;
  width: 100%;
}

.collapse {
  width: var(--sidebar-width-collapse);

  .switchers {
    display: flex;
    flex-direction: column;
    align-items: center;
  }

  .link-text {
    display: none;
    opacity: 0;
  }
}

.language-switcher {
  margin-left: 4px;
}

.button {
  position: absolute;
  right: -32px;
  bottom: 32px;
}
