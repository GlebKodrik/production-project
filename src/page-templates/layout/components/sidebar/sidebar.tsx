import React, { useState } from 'react';
import cls from 'classnames';
import { ThemeSwitcher } from 'feature/theme-switcher';
import { LanguageSwitcher } from 'feature/language-switcher';
import { Button } from 'shared-components/button';
import { SidebarLinks } from './components/sidebar-link';
import styles from './sidebar.module.scss';

export const Sidebar = () => {
  const [collapse, setCollapse] = useState(false);

  const onToggleCollapse = () => {
    setCollapse((prev) => !prev);
  };

  const renderButtonCollapse = () => (collapse ? '>' : '<');

  return (
    <aside
      data-testid="sidebar"
      className={cls(styles.sidebar, { [styles.collapse]: collapse })}
    >
      <div className={styles.linkWrapper}>
        <SidebarLinks />
      </div>

      <Button
        type="button"
        variant="square"
        onClick={onToggleCollapse}
        className={styles.button}
        size="large"
      >
        { renderButtonCollapse() }
      </Button>

      <div className={cls(styles.switchers)}>
        <ThemeSwitcher />
        <LanguageSwitcher
          className={styles.languageSwitcher}
          short={collapse}
        />
      </div>
    </aside>
  );
};
