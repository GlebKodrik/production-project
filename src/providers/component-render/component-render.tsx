import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { TComponentRenderProps } from './types';
import { TReduxStateScheme } from '../../redux/types/redux-state-scheme';
import { createStore } from '../../redux/root-store/root-store';

export const componentRender = ({ component, options = {}, initialState }: TComponentRenderProps) => {
  const {
    route = '/',
  } = options;

  const store = createStore(initialState as TReduxStateScheme);

  return render(
    <Provider store={store}>
      <MemoryRouter initialEntries={[route]}>
        {component}
      </MemoryRouter>
    </Provider>,
  );
};
