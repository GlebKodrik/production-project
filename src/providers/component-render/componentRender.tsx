import { ReactNode } from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { TComponentRenderOptions } from './types';

export const componentRender = (component: ReactNode, options: TComponentRenderOptions = {}) => {
  const {
    route = '/',
  } = options;

  return render(
    <MemoryRouter initialEntries={[route]}>
      {component}
    </MemoryRouter>,
  );
};
