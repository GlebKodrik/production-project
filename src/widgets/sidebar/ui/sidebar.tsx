import React, { useState } from 'react';
import cls from '../../../shared/lib/class-names';
import styles from './sidebar.module.scss';
import { ThemeSwitcher } from '../../../shared/ui/theme-switcher';
import {
  LanguageSwitcher,
} from '../../../shared/ui/language-switcher/language-switcher';

export const Sidebar = () => {
  const [collapse, setCollapse] = useState(false);

  const onToggleCollapse = () => {
    setCollapse((prev) => !prev);
  };

  return (
    <div
      data-testid="sidebar"
      className={cls(styles.sidebar, { [styles.collapse]: collapse })}
    >
      <button type="button" onClick={onToggleCollapse}>...</button>
      <div className={cls(styles.switchers)}>
        <ThemeSwitcher />
        <LanguageSwitcher className={styles.languageSwitcher} />
      </div>
    </div>
  );
};
