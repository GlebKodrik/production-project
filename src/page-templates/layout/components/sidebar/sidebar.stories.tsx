import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import { Sidebar } from './sidebar';

import { ThemeDecorator } from '../../../../configs/decorators/theme-decorator';
import { THEMES } from '../../../../constants/themes';

export default {
  title: 'widgets/Sidebar',
  component: Sidebar,
} as ComponentMeta<typeof Sidebar>;

const Template: ComponentStory<typeof Sidebar> = () => <Sidebar />;

export const SidebarWithThemeDark = Template.bind({});
SidebarWithThemeDark.args = {};
SidebarWithThemeDark.decorators = [ThemeDecorator(THEMES.DARK)];

export const SidebarWithThemeLight = Template.bind({});
SidebarWithThemeLight.args = {};
