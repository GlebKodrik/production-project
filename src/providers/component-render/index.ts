import { componentRender } from './component-render';

export { componentRender };
