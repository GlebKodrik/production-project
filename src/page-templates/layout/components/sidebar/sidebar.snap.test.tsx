import React from 'react';
import { componentRender } from 'providers/component-render';
import { Sidebar } from './sidebar';

describe('Snapshots of Sidebar component', () => {
  test('Snapshots of Sidebar component', () => {
    const { asFragment } = componentRender({ component: <Sidebar /> });
    expect(asFragment()).toMatchSnapshot();
  });
});
