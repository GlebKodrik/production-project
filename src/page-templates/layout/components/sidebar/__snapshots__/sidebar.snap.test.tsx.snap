// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshots of Sidebar component Snapshots of Sidebar component 1`] = `
<DocumentFragment>
  <aside
    class="sidebar"
    data-testid="sidebar"
  >
    <div
      class="linkWrapper"
    >
      <nav>
        <ul>
          <li>
            <a
              class="link primary"
              href="/"
            >
              <div
                class="item"
              >
                <div />
                <span
                  class="linkText"
                >
                  sidebar.links.mainLink
                </span>
              </div>
            </a>
          </li>
          <li>
            <a
              class="link primary"
              href="/about"
            >
              <div
                class="item"
              >
                <div />
                <span
                  class="linkText"
                >
                  sidebar.links.aboutLink
                </span>
              </div>
            </a>
          </li>
        </ul>
      </nav>
    </div>
    <button
      class="button variant-square color-primary size-large button"
      type="button"
    >
      &lt;
    </button>
    <div
      class="switchers"
    >
      <button
        class="button variant-clear color-primary size-medium"
        type="button"
      >
        <div />
      </button>
      <button
        class="button variant-clear color-primary size-medium languageSwitcher"
        type="button"
      >
        sidebar.language
      </button>
    </div>
  </aside>
</DocumentFragment>
`;
